core/iwasm/interpreter/wasm_runtime.c

```diff
@@ -326,12 +340,19 @@ memory_instantiate(WASMModuleInstance *module_inst, WASMModuleInstance *parent,
     memory->max_page_count = max_page_count;
     memory->memory_data_size = memory_data_size;
 
-    memory->heap_data = memory->memory_data + heap_offset;
-    memory->heap_data_end = memory->heap_data + heap_size;
-    memory->memory_data_end = memory->memory_data + memory_data_size;
+    if (is_main_memory) {
+        memory->heap_data = memory->memory_data + heap_offset;
+        memory->heap_data_end = memory->heap_data + heap_size;
+        memory->memory_data_end = memory->memory_data + memory_data_size;
+    }
+    else {
+        memory->heap_data = NULL;
+        memory->heap_data_end = NULL;
+        memory->memory_data_end = NULL;
```

Comment:
By default they are zeroed after memory structure is allocated, can remove the else branch?